src/statement.d

```diff
@@ -2478,15 +2478,21 @@ public:
                  *   for (T[] tmp = a[], size_t key = tmp.length; key--; )
                  *   { T value = tmp[k]; body }
                  */
-                Identifier id = Identifier.generateId("__r");
+                auto id = Identifier.generateId("__r");
                 auto ie = new ExpInitializer(loc, new SliceExp(loc, aggr, null, null));
                 VarDeclaration tmp;
                 if (aggr.op == TOKarrayliteral &&
                     !((*parameters)[dim - 1].storageClass & STCref))
                 {
                     auto ale = cast(ArrayLiteralExp)aggr;
                     size_t edim = ale.elements ? ale.elements.dim : 0;
-                    aggr.type = tab.nextOf().sarrayOf(edim);
+                    auto telem = (*parameters)[dim - 1].type;
+
+                    // Bugzilla 12936: if telem has been specified explicitly,
+                    // converting array literal elements to telem might make it @nogc.
+                    aggr = aggr.implicitCastTo(sc, telem.sarrayOf(edim));
+                    if (aggr.op == TOKerror)
+                        goto Lerror2;
```

Comment:
Essentially, no. The target of element type `telem` is either:
- explicitly specified by user, If it doesn't match `aggr` elements, it should have been error already.
- inferred from `aggr` elements type. If `implicitCastTo` call will cause a new error, it's a bug in `dcast.d`.